pkg/waiter/manager.go

```diff
@@ -0,0 +1,194 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+const (
+	maxIDCollisions = 5
+)
+
+var (
+	// ErrTooManyCollisions the ID generator produced too many non-unique IDs.
+	ErrTooManyCollisions = errors.New("too many id collisions")
+
+	// ErrManagerShutdown manager has been shutdown and therefore cannot process requests.
+	ErrManagerShutdown = errors.New("manager is shutdown")
+)
+
+type response[T any] struct {
+	id   string
+	data T
+	err  error
+}
+
+// Manager builds waiters and delivers messages to waiters from async publishers.
+//
+//go:generate mockgen-wrapper
+type Manager[T any] interface {
+	// Start spawns a goroutine that will run forever or until ctx.Done() delivering
+	// messages to waiters.
+	Start(ctx context.Context)
+
+	// Send sends data and err to the waiter with the provided id.
```

Comment:
would it be better to return an error instead of just no-oping?

Reply:
> Can we make it clear only the first call is recognized and subsequent calls are essentially no-ops?

Will add a comment.

> would it be better to return an error instead of just no-oping?

Possibly, to do so would require adding an additional locked section to Send() which I'd prefer to avoid. I'm not clear how the error would benefit the sender. The no-op could be logged, but couldn't be resent. If the waiter (originator of the request) doesn't get the desired response (perhaps due to timeout) it will be able to take the appropriate action (retry for example) which I believe is where that action belongs. I'll add a debug log to the Manager when a response is received for a non-existent waiter.

I'd prefer to keep the async 'spirit' of the package in tact, believe that will enable future use cases where we wouldn't be able to get that confirmation. If you're OK with it.